crates/infra/utils/src/codegen/runtime.rs

```diff
@@ -1,81 +1,75 @@
+        input_dir: impl Into<PathBuf>,
+        model: impl Serialize,
+    ) -> Result<()> {
+        let tera = TeraWrapper::new(input_dir)?;
         let context = tera::Context::from_serialize(model)?;
 
-        for template_path in self.tera.find_all_templates()? {
+        for template_path in tera.find_all_templates()? {
             let stub_path = Self::get_stub_path(&template_path).with_extension("");
-            let output = self.tera.render(&template_path, &context)?;
+            let output = tera.render(&template_path, &context)?;
 
-            self.fs.write_file(&stub_path, output)?;
+            fs.write_file_formatted(&stub_path, output)?;
         }
 
         Ok(())
     }
 
     pub fn render_product(
-        mut self,
-        model: impl Serialize,
+        fs: &mut CodegenFileSystem,
+        input_dir: impl Into<PathBuf>,
         output_dir: impl AsRef<Path>,
-    ) -> Result<CodegenFileSystem> {
-        let context = tera::Context::from_serialize(model)?;
+        model: impl Serialize,
+    ) -> Result<()> {
+        let input_dir = input_dir.into();
         let output_dir = output_dir.as_ref();
 
+        let tera = TeraWrapper::new(&input_dir)?;
+        let context = tera::Context::from_serialize(model)?;
+
         let mut handled = HashSet::new();
 
-        for template_path in self.tera.find_all_templates()? {
+        for template_path in tera.find_all_templates()? {
             let stub_path = Self::get_stub_path(&template_path).with_extension("");
-            let rendered_path = stub_path.replace_prefix(&self.input_dir, output_dir);
-            let rendered = self.tera.render(&template_path, &context)?;
+            let rendered_path = stub_path.replace_prefix(&input_dir, output_dir);
+            let rendered = tera.render(&template_path, &context)?;
 
-            self.fs.write_file(&rendered_path, rendered)?;
+            fs.write_file_formatted(&rendered_path, rendered)?;
 
             assert!(handled.insert(template_path));
             assert!(handled.insert(stub_path));
         }
 
-        for source_path in FileWalker::from_directory(&self.input_dir)
+        for source_path in FileWalker::from_directory(&input_dir)
             .find_all()?
             .filter(|source_path| !handled.contains(source_path))
         {
-            let output_path = source_path.replace_prefix(&self.input_dir, output_dir);
-
-            // Preserve the source of otherwise-generated files:
-            if source_path.generated_dir().is_ok() {
-                self.fs.mark_generated_file(output_path)?;
-            } else {
-                self.fs.copy_file(source_path, output_path)?;
-            }
+            let contents = source_path.read_to_string()?;
+            let output_path = source_path.replace_prefix(&input_dir, output_dir);
+
+            ensure!(
+                source_path.generated_dir().is_err(),
+                "Source file is generated: {source_path:?}"
+            );
```

Comment:
I find this hard to follow. Maybe improve the error message?

```suggestion
            ensure!(
                source_path.generated_dir().is_err(),
                "Source file should not be inside a generated directory: {source_path:?}"
            );
```

Maybe even adding another extension function so the condition can be written as `source_path.is_not_generated()`.

Reply:
Done.